src/webgpu/shader/validation/parse/shadow_builtins.spec.ts

```diff
@@ -0,0 +1,1002 @@
+  },
+  pack4xU8Clamp: {
+    keyword: `pack4xU8Clamp`,
+    src: `_ = pack4xU8Clamp(vec4(1, 2, 3, 4));`,
+  },
+  pow: {
+    keyword: `pow`,
+    src: `_ = pow(1, 2);`,
+  },
+  quantizeToF16: {
+    keyword: `quantizeToF16`,
+    src: `_ = quantizeToF16(1.2);`,
+  },
+  radians: {
+    keyword: `radians`,
+    src: `_ = radians(1.2);`,
+  },
+  reflect: {
+    keyword: `reflect`,
+    src: `_ = reflect(vec2(1, 2), vec2(3, 4));`,
+  },
+  refract: {
+    keyword: `refract`,
+    src: `_ = refract(vec2(1, 1), vec2(2, 2), 3);`,
+  },
+  reverseBits: {
+    keyword: `reverseBits`,
+    src: `_ = reverseBits(1);`,
+  },
+  round: {
+    keyword: `round`,
+    src: `_ = round(1.2);`,
+  },
+  saturate: {
+    keyword: `saturate`,
+    src: `_ = saturate(1);`,
+  },
+  select: {
+    keyword: `select`,
+    src: `_ = select(1, 2, false);`,
+  },
+  sign: {
+    keyword: `sign`,
+    src: `_ = sign(1);`,
+  },
+  sin: {
+    keyword: `sin`,
+    src: `_ = sin(2);`,
+  },
+  sinh: {
+    keyword: `sinh`,
+    src: `_ = sinh(3);`,
+  },
+  smoothstep: {
+    keyword: `smoothstep`,
+    src: `_ = smoothstep(1, 2, 3);`,
+  },
+  sqrt: {
+    keyword: `sqrt`,
+    src: `_ = sqrt(24);`,
+  },
+  step: {
+    keyword: `step`,
+    src: `_ = step(4, 5);`,
+  },
+  tan: {
+    keyword: `tan`,
+    src: `_ = tan(2);`,
+  },
+  tanh: {
+    keyword: `tanh`,
+    src: `_ = tanh(2);`,
+  },
+  transpose: {
+    keyword: `transpose`,
+    src: `_ = transpose(mat2x2(1, 2, 3, 4));`,
+  },
+  trunc: {
+    keyword: `trunc`,
+    src: `_ = trunc(2);`,
+  },
+  u32: {
+    keyword: `u32`,
+    src: `_ = u32(1i);`,
+  },
+  unpack2x16snorm: {
+    keyword: `unpack2x16snorm`,
+    src: `_ = unpack2x16snorm(2);`,
+  },
+  unpack2x16unorm: {
+    keyword: `unpack2x16unorm`,
+    src: `_ = unpack2x16unorm(2);`,
+  },
+  unpack2x16float: {
+    keyword: `unpack2x16float`,
+    src: `_ = unpack2x16float(2);`,
+  },
+  unpack4x8snorm: {
+    keyword: `unpack4x8snorm`,
+    src: `_ = unpack4x8snorm(4);`,
+  },
+  unpack4x8unorm: {
+    keyword: `unpack4x8unorm`,
+    src: `_ = unpack4x8unorm(4);`,
+  },
+  unpack4xI8: {
+    keyword: `unpack4xI8`,
+    src: `_ = unpack4xI8(4);`,
+  },
+  unpack4xU8: {
+    keyword: `unpack4xU8`,
+    src: `_ = unpack4xU8(4);`,
+  },
+  vec2_templated: {
+    keyword: `vec2`,
+    src: `_ = vec2<f32>(1, 2);`,
+  },
+  vec2: {
+    keyword: `vec2`,
+    src: `_ = vec2(1, 2);`,
+  },
+  vec3_templated: {
+    keyword: `vec3`,
+    src: `_ = vec3<f32>(1, 2, 3);`,
+  },
+  vec3: {
+    keyword: `vec3`,
+    src: `_ = vec3(1, 2, 3);`,
+  },
+  vec4_templated: {
+    keyword: `vec4`,
+    src: `_ = vec4<f32>(1, 2, 3, 4);`,
+  },
+  vec4: {
+    keyword: `vec4`,
+    src: `_ = vec4(1, 2, 3, 4);`,
+  },
+};
+
+g.test('shadow_hides_builtin')
+  .desc(`Test that shadows hide buitins.`)
+  .params(u =>
+    u
+      .combine('builtin', keysOf(kTests))
+      .combine('inject', ['none', 'function', 'sibling', 'module'])
```

Comment:
```suggestion
      .combine('inject', ['none', 'function', 'sibling', 'module'] as const)
```

The `as const` will make TypeScript treat `t.params.inject` as a string-type of `'none' | 'function' | 'sibling' | 'module'`, and typos will be caught at compile time instead of runtime.

Reply:
done